File: src/api/Domain.fs
namespace GW2Reroll.Api

module Domain =
    type Profession = 
        | Warrior
        | Guardian
        | Revenant
        | Ranger
        | Engineer
        | Thief
        | Elementalist
        | Necromancer
        | Mesmer
    
    type Race =
        | Human
        | Charr
        | Norn
        | Asura
        | Sylvari

    type Gender =
        | Male
        | Female
    
    type Character = Race * Profession * Gender
    
    let mapToProfession profession =
        match profession with
        | "Warrior" -> Some Warrior
        | "Guardian" -> Some Guardian
        | "Revenant" -> Some Revenant
        | "Ranger" -> Some Ranger
        | "Engineer" -> Some Engineer
        | "Thief" -> Some Thief
        | "Elementalist" -> Some Elementalist
        | "Necromancer" -> Some Necromancer
        | "Mesmer" -> Some Mesmer
        | _ -> None
    
    let mapToRace race =
        match race with
        | "Human" -> Some Human
        | "Charr" -> Some Charr
        | "Norn" -> Some Norn
        | "Asura" -> Some Asura
        | "Sylvari" -> Some Sylvari
        | _ -> None
    
    let mapToGender gender =
        match gender with
        | "Male" -> Some Male
        | "Female" -> Some Female
        | _ -> None
    

    let createCharacter (race, profession, gender): Character =
        (race, profession, gender)

File: src/shared/domain.ts
type Gender = "Male" | "Female";
type Race = "Human" | "Charr" | "Norn" | "Asura" | "Sylvari";
type Profession =
  | "Warrior"
  | "Guardian"
  | "Revenant"
  | "Ranger"
  | "Engineer"
  | "Thief"
  | "Elementalist"
  | "Necromancer"
  | "Mesmer";

export type Character = {
  name: string;
  gender: Gender;
  race: Race;
  profession: Profession;
  level: number;
};

export function recommend(characterData: Character[]): string {
  const profs = new Array<Profession>(
    "Warrior",
    "Guardian",
    "Revenant",
    "Ranger",
    "Thief",
    "Engineer",
    "Elementalist",
    "Necromancer",
    "Mesmer"
  );

  if (characterData.length === 0) {
    return "EMPTY";
  }

  let diff = profs.filter(
    (p) => !characterData.map((c) => c.profession).includes(p)
  );

  if (diff.length > 0) {
    return JSON.stringify(diff);
  }

  return "NAY";
}
